@alloy => New Show schema, and includes ExternalPartner

mzikherman opened the pull request:
This includes the Galaxy commits from https://github.com/artsy/metaphysics/pull/422

Idk if this is the right approach, but we can now get the partner name from a show like this (which works for Artsy or Galaxy partners):

<img width="823" alt="screen shot 2016-10-05 at 10 28 13 pm" src="https://cloud.githubusercontent.com/assets/1457859/19138532/087435a8-8b4b-11e6-9002-d32bd233d174.png">

However, this breaks the regular approach:

<img width="960" alt="screen shot 2016-10-05 at 10 28 52 pm" src="https://cloud.githubusercontent.com/assets/1457859/19138544/20eb7394-8b4b-11e6-9f52-cd5d83deff7d.png">

I don't mind updating Force (and other clients), but there might be lots of places that query partner attributes (from a show) in the original style. Is this the right approach? Or something more graceful?

alloy commented:
Right, hah, this is what I was referring to in #422, I had not yet read this PR.

Here’s an example of a Relay query for the artist view:

```graphql
query Routes($id_0:String!,$sort_1:ArtworkSorts!,$filter_2:[ArtistArtworksFilters]!,$filter_3:[ArtistArtworksFilters]!) {
  artist(id:$id_0) {
    __id,
    ...Fe
  }
}
fragment F0 on Artist {
  _id,
  id,
  name,
  nationality,
  birthday,
  counts {
    follows
  },
  __id
}
fragment F1 on Artist {
  href,
  name,
  counts {
    for_sale_artworks,
    artworks
  },
  image {
    _urlZizbi:url(version:"large")
  },
  __id
}
fragment F2 on Artist {
  __id,
  ...F1
}
fragment F3 on Article {
  thumbnail_title,
  href,
  author {
    name,
    __id
  },
  thumbnail_image {
    _urlZizbi:url(version:"large")
  },
  __id
}
fragment F4 on Article {
  __id,
  ...F3
}
fragment F5 on Artist {
  bio,
  blurb,
  __id
}
fragment F6 on Artist {
  has_metadata,
  _artists2sHGrU:artists(size:16) {
    __id,
    ...F2
  },
  articles {
    __id,
    ...F4
  },
  __id,
  ...F5
}
fragment F7 on PartnerShow {
  kind,
  name,
  exhibition_period,
  status_update,
  status,
  partner {
    name,
    __id
  },
  location {
    city,
    __id
  },
  __id
}
fragment F8 on PartnerShow {
  href,
  cover_image {
    _urlZizbi:url(version:"large")
  },
  __id,
  ...F7
}
fragment F9 on PartnerShow {
  __id,
  ...F8
}
fragment Fa on Artist {
  _partner_shows3spJoM:partner_shows(status:"running") {
    __id,
    ...F9
  },
  _partner_shows36OKdV:partner_shows(status:"upcoming") {
    __id,
    ...F9
  },
  _partner_shows2wrloS:partner_shows(status:"closed",size:20) {
    __id,
    ...F9
  },
  __id
}
fragment Fb on Artwork {
  title,
  date,
  sale_message,
  is_in_auction,
  image {
    _urlZizbi:url(version:"large"),
    aspect_ratio
  },
  artist {
    name,
    __id
  },
  partner {
    name,
    __id
  },
  href,
  __id
}
fragment Fc on Artwork {
  __id,
  image {
    aspect_ratio
  },
  ...Fb
}
fragment Fd on Artist {
  counts {
    artworks,
    for_sale_artworks
  },
  _artworks14j4Fu:artworks(sort:$sort_1,filter:$filter_2,size:10) {
    __id,
    ...Fc
  },
  _artworks1rDKOZ:artworks(sort:$sort_1,filter:$filter_3,size:10) {
    __id,
    ...Fc
  },
  __id
}
fragment Fe on Artist {
  _id,
  id,
  has_metadata,
  counts {
    artworks,
    partner_shows,
    related_artists,
    articles
  },
  __id,
  ...F0,
  ...F6,
  ...Fa,
  ...Fd
}
```

And some variables:

```json
{
  id_0: "ed-ruscha",
  sort_1: "partner_updated_at_desc",
  filter_2: "IS_FOR_SALE",
  filter_3: "IS_NOT_FOR_SALE"
}
```

I was hoping that maybe Relay always inserted the `... on FieldType { }` part, but it doesn’t seem to do so, so I’m pretty sure this query is going to break on this PR. Can you verify that?

alloy commented:
I can’t come up with any other name for the `partner` field (if you know of one, please do tell), so right now I see these options:

* Add a new field specifically for data coming from galaxy (lame) and we’d fix it when/if we’d move the schema to Gravity (because the change in URL makes it effectively a new endpoint version).

* Add a new `show` field wherever `partner_show` is used that duplicates `partner_show` pretty much, except that the `partner` field is a union and deprecate the `partner_show` fields.

  I actually think it cleans up the API a bit, unless a ‘show’ is too ambiguous for reasons I’m not seeing atm, we’d just have to be very deliberate in reviewing if any other fields might also need any backwards incompatible changes, because we can’t easily keep doing this.

mzikherman commented:
> Add a new field specifically for data coming from galaxy (lame) and we’d fix it when/if we’d move the schema to Gravity (because the change in URL makes it effectively a new endpoint version).

Do you mean adding `galaxy_partner` to the schema for the show (rather than union'ing the existing partner node), and then the places that want to use it would have to specify both `partner` and `galaxy_partner`? That's reasonable to me since it won't break any existing stuff, and (as of now), there is one specific use-case/place we want to display it on the front-end.

> Add a new show field wherever partner_show is used that duplicates partner_show pretty much, except that the partner field is a union and deprecate the partner_show fields.

I don't exactly follow?

mzikherman commented:
Also as expected, that Relay query now fails with the same error: `"Cannot query field \"name\" on type \"PartnerTypes\"`

mzikherman commented on schema/artist/index.js:
```diff
@@ -321,7 +321,6 @@ const ArtistType = new GraphQLObjectType({
             gravity('related/shows', {
               artist_id: id,
               sort: '-end_at',
-              displayable: true,
```
These can be killed, it was obviated as a param a client needs to pass in.

mzikherman commented on schema/galaxy_partner.js:
```diff
@@ -0,0 +1,35 @@
+import galaxy from '../lib/loaders/galaxy';
```
barebones schema for a response, looks like:

`{"id":41001,"name":"Vladimir Sokolov Studio Gallery","_links":{"self":{"href":"https://artsy-galaxy-staging.herokuapp.com/galleries/41001"}}}`

( I will most likely add some location stuff in the immediate future too)

mzikherman commented on schema/partner_show.js:
```diff
@@ -100,6 +101,14 @@ const PartnerShowType = new GraphQLObjectType({
       type: new GraphQLList(Artist.type),
       resolve: ({ artists }) => artists,
     },
+    galaxy_partner: {
```
just make a new node for a galaxy partner. The CV/exhibition history fragment would then look like:

```
partner {
  name
}
galaxy_partner {
  name
}
```

I can adjust the template (or add a helper) to pick the exist-y one.

mzikherman commented on schema/partner_show.js:
```diff
@@ -269,7 +278,7 @@ const PartnerShow = {
     return gravity(`show/${id}`)
       .then(show => {
-        if (!show.displayable) return new Error('Show Not Found');
+        if (!show.displayable && !show.is_reference) return new Error('Show Not Found');
```
We sort of use `displayable` to mean something ambiguous. It essentially means if the show has a standalone page on the site, which ref shows do not. I'm not sure if this was the best place to put this `Error` checking. I think the standalone show page on the site (powered by MP) should do this check, since it's non-standard.

mzikherman commented:
Ok, updated, I think this is mergable.

I copied over the specs and schema for `PartnerShow` into `Show` and modified the `partner` node to be a union type of `GalaxyPartner` or `Partner` (+ a spec for that).

alloy commented:
Yup, this looks good to me, code wise 👍

The one thing I worry about is the type name, ‘galaxy’ only refers to the underlying data source, it’s not descriptive to readers of the schema or those unaware of what the Galaxy service serves exactly. What’s the terminology used in the business to refer to such partners? Something like `UnaffiliatedPartner`?

mzikherman commented:
Good point, it refers to the service and not the function that the service serves.

I'm often torn about such naming in code, sometimes since the functionality is so tied to the backing service, and would require a total change if the service ever got replaced, I've been feeling that tying the name to the service makes the most sense and is explicit, vs something like `Unaffiliated Partner` which abstracts that.

I'm not sure the standard naming for such a gallery, that sounds good to me 👍

alloy commented:
I could imagine us changing the complete service, but clients shouldn’t _need_ to know about that. So the loader and whatever other internal code should totally refer to `galaxy`, but in the public schema I think those details should be abstracted away.

Judging from the [README](https://github.com/artsy/galaxy-api/blob/master/README.md#galaxy-api), they appear to be referred to as ‘external partners’. I can’t say I find that terribly descriptive, but, on the other hand, ‘unaffiliated partner’ is kind of an oxymoron so best stick to the nomenclature at use in the business ¯\_(ツ)_/¯

mzikherman commented:
I see, that makes sense to me! I'll give the renaming a shot.

mzikherman commented:
Renamed to `ExternalPartner`

alloy commented on schema/index.js:
```diff
@@ -46,9 +46,9 @@ const schema = new GraphQLSchema({
+      external_partner: ExternalPartner,
       fair: Fair,
       fairs: Fairs,
-      galaxy_partner: GalaxyPartner,
```
FYI we don’t use the `partner` root field in Eigen atm, so it will be easy to convert the `partner` field to a union.

alloy commented:
Looking 👌
